src/lib/optimizer/enumerate_ccp.cpp

```diff
@@ -0,0 +1,174 @@
+
+  for (const auto& subset : neighborhood_subsets) {
+    csgs.emplace_back(subset | vertex_set);
+  }
+
+  for (const auto& subset : neighborhood_subsets) {
+    _enumerate_csg_recursive(csgs, subset | vertex_set, extended_exclusion_set);
+  }
+}
+
+void EnumerateCcp::_enumerate_cmp(const JoinGraphVertexSet& vertex_set) {
+  const auto exclusion_set = _exclusion_set(vertex_set.find_first()) | vertex_set;
+  const auto neighborhood = _neighborhood(vertex_set, exclusion_set);
+
+  if (neighborhood.none()) return;
+
+  std::vector<size_t> reverse_vertex_indices;
+  auto current_vertex_idx = neighborhood.find_first();
+
+  do {
+    reverse_vertex_indices.emplace_back(current_vertex_idx);
+  } while ((current_vertex_idx = neighborhood.find_next(current_vertex_idx)) != JoinGraphVertexSet::npos);
+
+  for (auto iter = reverse_vertex_indices.rbegin(); iter != reverse_vertex_indices.rend(); ++iter) {
+    auto cmp_vertex_set = JoinGraphVertexSet(_num_vertices);
+    cmp_vertex_set.set(*iter);
+
+    _csg_cmp_pairs.emplace_back(std::make_pair(vertex_set, cmp_vertex_set));
+
+    const auto extended_exclusion_set = exclusion_set | (_exclusion_set(*iter) & neighborhood);
+
+    std::vector<JoinGraphVertexSet> csgs;
+    _enumerate_csg_recursive(csgs, cmp_vertex_set, extended_exclusion_set);
+
+    for (const auto& csg : csgs) {
+      _csg_cmp_pairs.emplace_back(std::make_pair(vertex_set, csg));
+    }
+  }
+}
+
+JoinGraphVertexSet EnumerateCcp::_exclusion_set(const size_t vertex_idx) const {
+  JoinGraphVertexSet exclusion_set(_num_vertices);
+  for (size_t exclusion_vertex_idx = 0; exclusion_vertex_idx < vertex_idx; ++exclusion_vertex_idx) {
+    exclusion_set.set(exclusion_vertex_idx);
+  }
+  return exclusion_set;
+}
+
+JoinGraphVertexSet EnumerateCcp::_neighborhood(const JoinGraphVertexSet& vertex_set,
+                                               const JoinGraphVertexSet& exclusion_set) const {
+  JoinGraphVertexSet neighborhood(_num_vertices);
+
+  for (auto current_vertex_idx = size_t{0}; current_vertex_idx < _num_vertices; ++current_vertex_idx) {
+    if (!vertex_set[current_vertex_idx]) continue;
+
+    neighborhood |= _vertex_neighborhoods[current_vertex_idx];
+  }
+
+  return neighborhood - exclusion_set - vertex_set;
+}
+
+JoinGraphVertexSet EnumerateCcp::_single_vertex_neighborhood(const size_t vertex_idx) const {
+  JoinGraphVertexSet neighbourhood(_num_vertices);
+  for (const auto& edge : _edges) {
+    if (vertex_idx == edge.first && vertex_idx != edge.second) {
+      neighbourhood.set(edge.second);
+    }
+    if (vertex_idx != edge.first && vertex_idx == edge.second) {
+      neighbourhood.set(edge.first);
+    }
+  }
+
+  return neighbourhood;
+}
+
+std::vector<JoinGraphVertexSet> EnumerateCcp::_non_empty_subsets(const JoinGraphVertexSet& vertex_set) const {
+  if (vertex_set.none()) return {};
+
+  std::vector<JoinGraphVertexSet> subsets;
+
+  const auto s = vertex_set.to_ulong();
+  auto s1 = s & -s;
```

Comment:
That's not intuitive to me. Could you add a comment?

Reply:
I added a comment that I cannot give a convincing explanation of this algorithm. I added an explanation of its result, though

```
  /**
   * Returns all proper subsets of `vertex_set`, e.g., for 01101 return {00001, 00100, 01000, 01001, 01100}
   *
   * The algorithm listed here is from stackoverflow, but I can't find the link to it anymore, hard as I tried.
   * Neither can I convincingly explain the bit-magic here, but it works nicely.
   */
```

Reply:
Also added ` // s1 is the current subset subset [sic]. s & -s initializes it to the least significant bit in s`

Reply:
Does it also return `00101`?

Reply:
yes, looks like I forgot that one